File: src/algorithms/quicksort-int.js
export function quicksortInt(array, start, end) {
  if (start >= end) {
    return;
  }
  let pivot = array[end];
  let left = 0;
  let right = 0;
  while (left + right < end - start) {
    let num = array[start + left];
    if (num < pivot) {
      left++;
    } else {
      array[start + left] = array[end - right - 1];
      array[end - right - 1] = pivot;
      array[end - right] = num;
      right++;
    }
  }
  quicksortInt(array, start, start + left - 1);
  quicksortInt(array, start + left + 1, end);
}

export function randomizeIntArray(array) {
  for (var i = 0, il = array.length; i < il; i++) {
    array[i] = ((Math.random() * 20000) | 0) - 10000;
  }
}

File: src/algorithms/quicksort-int.c
#include <stdlib.h>
#include <time.h>

void quicksort_int(int *array, int start, int end) {
  if (start >= end) {
    return;
  }
  int pivot = array[end];
  int left = 0;
  int right = 0;
  while (left + right < end - start) {
    int num = array[start + left];
    if (num < pivot) {
      left++;
    } else {
      array[start + left] = array[end - right - 1];
      array[end - right - 1] = pivot;
      array[end - right] = num;
      right++;
    }
  }
  quicksort_int(array, start, start + left - 1);
  quicksort_int(array, start + left + 1, end);
}

void randomize_int_array(int *array) {
  srand(time(0));

  int l = sizeof(array) / sizeof(array[0]);
  for (int i = 0; i < l; i++) {
    array[i] = rand();
  }
}
